artemis-server/src/main/java/org/apache/activemq/artemis/core/postoffice/impl/SimpleAddressManager.java

```diff
@@ -100,6 +102,8 @@ public boolean addBinding(final Binding binding) throws Exception {
       if (nameMap.putIfAbsent(binding.getUniqueName(), bindingAddressPair) != null) {
          throw ActiveMQMessageBundle.BUNDLE.bindingAlreadyExists(binding);
       }
+      directBindingMap.computeIfAbsent(binding.getAddress(), (unused) -> new ArrayList<>())
+              .add(binding);
```

Comment:
you are using ConcurrentHashMap on  the directBindingsMap, however you're using a non Concurrent list on directBindingsMap. 

I'm wondering if we shouldn't use something more protective in multi-thread here. 

I understand it's unlikely the broker to remove or add a queue on the same address from multiple threads.. but it's not impossible. (say an application with a MultiCast address and applications creating and deleting subscriptions from multiple places).


it would be a rare race, but difficult to track if it ever happened. Shouldn't we use something more protective here.. perhaps CopyOnWrite or anything else?

Reply:
Good point, definitely should be thread-safe. Changed to a COW ArrayList.

Reply:
Using a copy on write structure would seem to be adding a significant amount of GC overhead as each new binding will copy the array and toss out the old instance.

Reply:
Now using an `ArrayList` given that the modifications are happening in a (synchronized) `compute` block.